src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -386,34 +410,156 @@ public override int GetHashCode()
 
         internal override void AddNullableTransforms(ArrayBuilder<byte> transforms)
         {
-            ElementTypeWithAnnotations.AddNullableTransforms(transforms);
+            var elementType = ElementTypeWithAnnotations;
+            while (elementType.Type is ArrayTypeSymbol arrayType)
+            {
+                elementType.AddNullableTransformShallow(transforms);
+                elementType = arrayType.ElementTypeWithAnnotations;
+            }
+
+            elementType.AddNullableTransforms(transforms);
         }
 
-        internal override bool ApplyNullableTransforms(byte defaultTransformFlag, ImmutableArray<byte> transforms, ref int position, out TypeSymbol result)
+        internal override TypeSymbol ApplyNullableTransforms(NullableTransformStream stream)
         {
-            TypeWithAnnotations oldElementType = ElementTypeWithAnnotations;
-            TypeWithAnnotations newElementType;
+            // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+            // recursive solution here. Need to take an iterative approach.
+            var builder = ArrayBuilder<(ArrayTypeSymbol array, byte? transform)>.GetInstance();
 
-            if (!oldElementType.ApplyNullableTransforms(defaultTransformFlag, transforms, ref position, out newElementType))
+            // First build up the state for each of the ArrayTypeSymbol in the chain of arrays.
+            var array = this;
+            builder.Push((this, stream.GetNextTransform()));
+            do
             {
-                result = this;
-                return false;
+                array = array.ElementType as ArrayTypeSymbol;
+                if (array is null)
+                {
+                    break;
+                }
+                else
+                {
+                    builder.Push((array, stream.GetNextTransform()));
+                }
             }
+            while (true);
+
+            // Create the initial return for the most nested array. 
+            var lastState = builder.Pop();
+            var lastElementType = apply(
+                lastState.array.ElementTypeWithAnnotations,
+                lastState.array.ElementTypeWithAnnotations.Type.ApplyNullableTransforms(stream),
+                lastState.transform);
+            var finalArray = lastState.array.WithElementType(lastElementType);
+
```

Comment:
I suspect this would be simpler if you started with the final element type (0 case) rather than the final array (1 case).